src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/Cas2ApplicationsController.kt

```diff
@@ -0,0 +1,157 @@
+  private val applicationsTransformer: ApplicationsTransformer,
+  private val objectMapper: ObjectMapper,
+  private val offenderService: OffenderService,
+  private val userService: UserService,
+) : Cas2ApiDelegate {
+  private val log = LoggerFactory.getLogger(this::class.java)
+
+  override fun cas2ApplicationsGet(): ResponseEntity<List<ApplicationSummary>> {
+    val user = userService.getUserForRequest()
+
+    val applications = applicationService.getAllApplicationsForUsername(
+      user
+        .deliusUsername,
+      ServiceName.cas2,
+    )
+
+    return ResponseEntity.ok(applications.map { getPersonDetailAndTransformToSummary(it, user) })
+  }
+
+  override fun cas2ApplicationsApplicationIdGet(applicationId: UUID):
+    ResponseEntity<Application> {
+    val user = userService.getUserForRequest()
+
+    val application = when (val applicationResult = applicationService.getApplicationForUsername(applicationId, user.deliusUsername)) {
+      is AuthorisableActionResult.NotFound -> null
+      is AuthorisableActionResult.Unauthorised -> throw ForbiddenProblem()
+      is AuthorisableActionResult.Success -> applicationResult.entity
+    }
+
+    if (application != null) {
+      return ResponseEntity.ok(getPersonDetailAndTransform(application, user))
+    }
+    throw NotFoundProblem(applicationId, "Application")
+  }
+
+  @Transactional
+  override fun cas2ApplicationsPost(body: NewApplication):
+    ResponseEntity<Application> {
+    val deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
+    val user = userService.getUserForRequest()
+
+    val personInfo = offenderService.getFullInfoForPersonOrThrow(body.crn, user)
+
+    val applicationResult = applicationService.createCas2Application(body.crn, user, deliusPrincipal.token.tokenValue)
+
+    val application = when (applicationResult) {
+      is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = applicationResult.message)
+      is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = applicationResult.validationMessages)
+      is ValidatableActionResult.ConflictError -> throw ConflictProblem(id = applicationResult.conflictingEntityId, conflictReason = applicationResult.message)
+      is ValidatableActionResult.Success -> applicationResult.entity
+    }
+
+    return ResponseEntity
+      .created(URI.create("/applications/${application.id}"))
+      .body(applicationsTransformer.transformJpaToApi(application, personInfo))
+  }
+
+  @Transactional
+  override fun cas2ApplicationsApplicationIdPut(
+    applicationId: UUID,
+    body:
+      UpdateApplication,
+  ): ResponseEntity<Application> {
+    val user = userService.getUserForRequest()
+
+    val serializedData = objectMapper.writeValueAsString(body.data)
+
+    val applicationResult = applicationService.updateCas2Application(applicationId = applicationId, data = serializedData, username = user.deliusUsername)
```

Comment:
I wonder if now might be the time to namespace the services too? So the `applicationService` in this case would be in `services/cas2/ApplicationService.kt`?

Reply:
yes good plan. I'd like to do this later refactoring step as this is already quite large?